File: 백준/Silver/1991. 트리 순회/트리 순회.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int n = Integer.parseInt(br.readLine());
        Tree tree = new Tree();

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            char parent = st.nextToken().charAt(0);
            char left = st.nextToken().charAt(0);
            char right = st.nextToken().charAt(0);

            tree.createNode(parent, left, right);
        }

        tree.preOrder(tree.root);
        System.out.println();
        tree.inOrder(tree.root);
        System.out.println();
        tree.postOrder(tree.root);

    }
}

class Node {
    char data;
    Node left;
    Node right;

    public Node(char data) {
        this.data = data;
    }

    public void insertLeft(char data) {
        this.left = (data == '.') ? null : new Node(data);
    }

    public void insertRight(char data) {
        this.right = (data == '.') ? null : new Node(data);
    }
}

class Tree{
    Node root;

    public void createNode(char parent, char left, char right) {
        if (root == null) {
            root = new Node(parent);
            root.insertLeft(left);
            root.insertRight(right);
        } else {
            //부모찾아서 거기에 left, right 삽입
            searchNode(root, parent, left, right);
        }
    }

    public void searchNode(Node node, char parent, char left, char right) {
        if (node == null) return;
        if (node.data == parent) {
            node.insertRight(right);
            node.insertLeft(left);
            return;
        }
        searchNode(node.left, parent, left, right);
        searchNode(node.right, parent, left, right);
    }

    public void preOrder(Node node) {
        if (node == null) return;
        System.out.print(node.data);
        preOrder(node.left);
        preOrder(node.right);
    }
    public void postOrder(Node node) {
        if (node == null) return;
        postOrder(node.left);
        postOrder(node.right);
        System.out.print(node.data);
    }
    public void inOrder(Node node) {
        if (node == null) return;
        inOrder(node.left);
        System.out.print(node.data);
        inOrder(node.right);
    }

}

File: 백준/Silver/1991. 트리 순회/트리 순회.c
#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>
#include <stdlib.h>

typedef struct node* treePointer;
typedef struct node {
	char data;
	treePointer leftChild;
	treePointer rightChild;
};

treePointer root = NULL;
void insert(char parent, char lchild, char rchild);
treePointer makeNode(char data);
void searchAndInsert(treePointer node, char parent, char left, char right);


void insert(char parent, char lchild, char rchild) {
	if (root == NULL) {
		treePointer newNode = makeNode(parent);
		if (lchild != '.') newNode->leftChild = makeNode(lchild);
		if (rchild != '.') newNode->rightChild = makeNode(rchild);
		root = newNode;
	}
	else {
		searchAndInsert(root, parent, lchild, rchild);
	}
}

treePointer makeNode(char data) {
	treePointer newNode = (treePointer)malloc(sizeof(struct node));
	newNode->data = data;
	newNode->leftChild = NULL;
	newNode->rightChild = NULL;
	return newNode;
}

void searchAndInsert(treePointer node, char parent, char left, char right) {
	if (node == NULL) return;
	if (node->data == parent) {
		if (left == '.') left = NULL;
		else node->leftChild = makeNode(left);

		if (right == '.') right = NULL;
		else node->rightChild = makeNode(right);
		return;
	}
	searchAndInsert(node->leftChild, parent, left, right);
	searchAndInsert(node->rightChild, parent, left, right);
}

void preorder(treePointer node) {
	if (node) {
		printf("%c", node->data);
		preorder(node->leftChild);
		preorder(node->rightChild);
	}
}

void inorder(treePointer node) {
	if (node) {
		inorder(node->leftChild);
		printf("%c", node->data);
		inorder(node->rightChild);
	}
}

void postorder(treePointer node) {
	if (node) {
		postorder(node->leftChild);
		postorder(node->rightChild);
		printf("%c", node->data);
	}
}

int main() {
	int n;
	scanf("%d", &n);
	char parent, lchild, rchild;
	for (int i = 0; i < n; i++) {
		scanf(" %c %c %c", &parent, &lchild, &rchild);
		insert(parent, lchild, rchild);
	}

	preorder(root);
	printf("\n");
	inorder(root);
	printf("\n");
	postorder(root);
}